gui/util/range_slider.py

```diff
@@ -0,0 +1,294 @@
+        self.prev_moving = None
+        self.bc_min = None
+        self.bc_max = None
+
+        # Variables initialized in methods
+        self.scale_min = None
+        self.scale_max = None
+        self.start_display_min = None
+        self.start_display_max = None
+        self.start_pos = None
+        self.display_min = None
+        self.display_max = None
+
+        if slider_range:
+            self.setRange(slider_range)
+        else:
+            self.setRange([0.0, 1.0, 0.01])
+        if values:
+            self.values = values
+        else:
+            self.values = [0.3, 0.6]
+
+    def emitRange(self):
+        if (self.old_scale_min != self.scale_min) or (self.old_scale_max != self.scale_max):
+            self.rangeChanged.emit(self.scale_min, self.scale_max)
+            self.old_scale_min = self.scale_min
+            self.old_scale_max = self.scale_max
+            # For debug purposes
+            # if False:
+            #     print("Range change:", self.scale_min, self.scale_max)
+
+    @property
+    def values(self):
+        """
+        Values of the range bar
+        :return: [start, end]
+        """
+        return [self.scale_min, self.scale_max]
+
+    @values.setter
+    def values(self, values):
+        """
+        Setter for values of the range bar
+        :param values: [start, end]
+        """
+        self.scale_min, self.scale_max = values
+        self.emitRange()
+        self.updateDisplayValues()
+        self.update()
+
+    def mouseMoveEvent(self, event):
+        size = self.rangeSliderSize()
+        diff = self.start_pos - self.getPos(event)
+        if self.moving == "min":
+            temp = self.start_display_min - diff
+            if (temp >= self.bar_width) and (temp <= self.start_display_max):
+                self.display_min = temp
+                if self.display_max < self.display_min:
+                    self.display_max = self.display_min
+        elif self.moving == "max":
+            temp = self.start_display_max - diff
+            if (temp >= self.start_display_min) and (temp < size - self.bar_width):
+                self.display_max = temp
+                if self.display_max < self.display_min:
+                    self.display_min = self.display_max
+        elif self.moving == "bar":
+            temp = self.start_display_min - diff
+            if (temp >= self.bar_width) and (
+                    temp < size - self.bar_width - (self.start_display_max - self.start_display_min)):
+                self.display_min = temp
+                self.display_max = self.start_display_max - diff
+
+        self.updateScaleValues()
+        if self.emit_while_moving:
+            self.emitRange()
+
+    def mousePressEvent(self, event):
+        pos = self.getPos(event)
+        if event.button() == QtCore.Qt.LeftButton:
+            if not self.collapsed:
+                if abs(self.display_min - 0.5 * self.bar_width - pos) <= (0.5 * self.bar_width):
```

Comment:
I don't understand why the distance being measured is from the left edge of the widget instead of from the center?